Make torch.cond work with retracing

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92646

We simplify the handling of branch submodules by only working with flattened input/output so that there is no need for adjusting in_spec and out_spec in the second round of tracing.

tugsbayasgalan committed:
Make torch.cond work with retracing

[ghstack-poisoned]

zhxchen17 commented on functorch/experimental/_cond.py:
```diff
@@ -40,8 +40,10 @@ def trace_cond(proxy_mode, func_overload, pred, true_fn, false_fn, operands):
 
-    true_graph = get_isolated_graphmodule(true_fn, operands, {})
-    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+    flattened_inputs, spec = pytree.tree_flatten(operands)
```
@tugsbayasgalan we don't need this line, since line 40 and 41 already guarantee you have a flattened list?

tugsbayasgalan replied:
Yeah you are right. Will update :)

tugsbayasgalan committed:
Update on "Make torch.cond work with retracing"


We simplify the handling of branch submodules by only working with flattened input/output so that there is no need for adjusting in_spec and out_spec in the second round of tracing.

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge

tugsbayasgalan commented:
@pytorchbot merge -f "Error seems not related"
